### DISCUSSION TITLE
Auth handler per connection

### DISCUSSION THREAD

**Discussion by jborean93** [encode/httpx#1473] [Q&A] [answered]
I'm currently trying out httpx as part of a WinRM based library which has a unique requirement for encryption over HTTP. The authentication is typically `Negotiate` authentication (Kerberos or NTLM) and while I've able to create a handler class that implements `httpx.Auth` that will successfully authenticate against the host I need to preserve the security context so I can encrypt the data sent to the client. This is a very basic mock up of what I currently have in place

```python
import httpx

class NegotiateAuth(httpx.Auth):

    def __init__(self, username, password):
        self.username = username
        self.password = None
        self.context = None

    def sync_auth_flow(self, request):
        response = yield request
        if response.status_code != 401:
            return

         self.context = MySecContext(self.username, self.password)
        out_token = self.context.step()
        while not self.context.complete:
            request.headers['Authorization'] = out_token

            # send the request with the auth token and get the response
            response = yield request

            # Feed the token back into the context until it's set up

    def wrap(self, data);
        return self.context.wrap(data)

    def unwrap(self, data):
        return self.context.unwrap(data)


class WinRMClient:

    def __init__(self):
        auth = NegotiateAuth('username', 'password')
        client = httpx.AsyncClient(auth=auth)

    async def send(self, data):
        # set up initial auth with blank request
        if not self.auth.context:
            await self.client.post('endpoint', content=b'')

        enc_data = self.client.auth.wrap(data)
        resp = await self.client.post('endpoint', content=enc_data)
        return self.client.auth.unwrap(resp.content)
```

This works fine if I've only ever got 1 connection being used in the pool as the context in the `auth` object always relates to the connection that was opened. Unfortunately as soon as I start sending multiple requests concurrently httpx will create a new connection from the pool starting the authentication process again for the new connection. Ultimately this will overwrite the context for the initial connection meaning I'm no longer able to wrap/unwrap data.

What I'm hoping to find out is a way to:

* Create a context for each connection
    * I've looked into subclassing `httpcore.AsyncConnectionPool` as a transport and overriding `_create_connection()` but it's not a public API so this is not ideal
    * Even though it's not public I think I could set up the auth context here and just use `setattr()` on the connection to store it but I really don't want to be messing around with these internals
* Intercepting something that prepares the requests and processes the response so that I can wrap/unwrap the data
    * I would need to be able to link the socket/connection that the request will be sent over to the security context for that connection
    * A lot of the `arequest()` functions I can see in some of the internal classes aren't really tied to an actual
    * I think I could override the `arequest()` method on the connection created by `_create_connection()` which transforms the data but like the above I prefer not to mess around with this stuff if I can avoid it

Do you have any ideas how I could potentially implement these requirements?

**Comment by florimondmanca** [comment, marked as answer]
Hey @davidbgk 

This is a very interesting scenario.

I believe the transport route is the most promising. The "wrap / unwrap" thingy you do there for example screams for "wrap the request / response stream" at the transport level. :)

But I wouldn't do subclassing. Instead I think you might be able to get something working by composing with an existing transport, and then wrapping things however is necessary to accommodate your auth scheme. You might not even need to modify the pool, and instead do the tracking of origins inside your own transport.

I can draft some code snippets to help you out with this intuition later if you'd like. :)

**Reply by jborean93** [reply]
Thanks, the actual WIP implemtation I have going is at https://github.com/jborean93/pypsrp/blob/asyncio/psrp/io/wsman.py. It’s very early stages and the code is rapidly changing as I go along. My hope is that this will use only public interfaces or at least be a driver for introducing more public and stable interfaces for the 1.0.0 release so happy to help wherever I can.


I also wanted to say, thank you very much for your help and work on this library. Seems like a very promising project and well put together.

**Reply by davidbgk** [reply]
@florimondmanca I'm not sure why you mentioned me, should I take a look at something in particular? Maybe I forgot about an issue I created on that topic? (if it's a typo, it's fine too 😅)

**Reply by florimondmanca** [reply]
@davidbgk :joy: Sorry! GitHub put you first in the @ suggestions, so I thought you were the OP! I was targeting @jborean93. :-)

@jborean93 Thanks for the kind words. :-) I think your case ties in with discussions we had previously about "stateful auth". In general, auth classes _shouldn't_ contain state. That's because we can't know if an `auth` will be used at the client level (via `Client(auth=...)`), in which case it gets reused across requests (so state may become corrupt), or at the request level (via `client.get(..., auth=...)`), in which case the auth isn't reused but then you can't store state either!

The general rule of thumb is: store state in the `transport` (which follows the same lifecycle as a `Client`, and can't be overridden on a per-request basis). Which is why I was hinting at writing a separate transport that would do the per-origin splitting.

We don't have very extensive tutorials on writing custom transports just yet, but there is some info here: https://www.python-httpx.org/advanced/#writing-custom-transports and detailed API documentation at https://www.encode.io/httpcore/.

Basically you'd end up with this:

```python
class WinRMClient:
    def __init__(self, client):
        self._client = client

    async def send(self, data):
        resp = await self._client.post('endpoint', content=data)
        return resp.content


@asynccontextmanager
async def open_win_rm_client():
    transport = httpx.HTTPTransport()  # To be released in 0.17.0, use `httpcore.AsyncConnectionPool` in the interim.
    transport = WinRMTransport(transport, username=..., password=...)
    async with httpx.AsyncClient(transport=transport) as client:
         await self.client.post('endpoint', content=b"")  # Set up initial auth with blank request.
        yield WinRMClient(client)
```

Note the separate `open_win_rm_client()` context manager, which takes care of initializing the client, and properly ensuring the proper cleanup of resources on exit, so you can use your client transparently like so:

```python
async with open_win_rm_client() as winrm:
    content = await winrm.send("test")
```

And I believe `WinRMTransport` could look like this:

```python
import httpcore
import httpx

class WinRMTransport(httpcore.AsyncHTTPTransport):
    def __init__(self, transport: httpcore.AsyncHTTPTransport, username: str, password: str):
        self._transport = transport
        self._username, self._password = username, password
        self._contexts = {}  # <-- Transport-level state

    async def arequest(self, method: bytes, url: tuple, headers: list=None, stream: httpcore.AsyncByteStream=None, ext: dict=None):
        # TODO: inspect `url` (which is a (scheme, host, port, path) tuple) and get-or-create
        # security contexts accordingly.
        context = self._contexts.get(...)
        if context is None:
            context = MySecContext(self._username, self._password)
            self._contexts[...] = context

        stream = self._wrap_request_stream(stream, context)
        response = await self._transport.arequest(method, url, headers, stream, ext)

        status_code, *_ = response
        if status_code != 401:
            return response

        out_token = self._context.step()
        while not self._context.complete:
            temp_headers = httpx.Headers(headers)  # Easier to work with than raw `headers` list.
            temp_headers["Authorization"] = out_token
            headers = temp_headers.raw

            # send the request with the auth token and get the response
            response = await self._transport.arequest(method, url, headers, stream, ext)

            # Feed the token back into the context until it's set up

        status_code, headers, stream, ext = response
        stream = self._wrap_response_stream(stream, context)

        return (status_code, headers, stream, ext)

    def _wrap_request_stream(self, stream, context):
        if stream is None:
            return None
        return MapStream(stream, mapfunc=context.wrap)

    def _wrap_response_stream(self, stream, context):
        if stream is None:
            return None
        return MapStream(stream, mapfunc=context.unwrap)


class MapStream(httpcore.AsyncByteStream):
    def __init__(self, stream: httpcore.AsyncByteStream, mapfunc: Callable):
        self._stream = stream
        self._mapfunc = mapfunc

    async def __aiter__(self):
        async for chunk in self._stream:
            yield self._mapfunc(chunk)

    async def aclose(self):
        await self._stream.aclose()
```

As you can see, you'd be able to A/ defer underlying operations (connection pooling, etc) to an opaque `transport`, B/ store security contexts on the transport, and manage them eg per `(scheme, host, port)` origin, C/ apply wrapping/unwrapping of data inside the transport, so that it's fully transparent to the client.

**Reply by jborean93** [reply]
Thanks for the info, the problem with your proposal is that it's assigning a unique context per scheme/host/port but in my case they are all the same. If I was to send 2 POST at the same time then httpx will create 2 connections and they both require require 2 different contexts. Both of these connections will be sending to `http://host:port/path` so I can't use the arguments to `arequest()` to select the proper context. I also can't rely on creating a new context for each request, this context must be use for the life of the connection as I need to send multiple POST requests.

Because the only time I can see that the connection and request comes into play is when [AsyncConnectionPool.arequest](https://github.com/encode/httpcore/blob/ed76a30277529c5687cbdb6e203d6dbd0d56833f/httpcore/_async/connection_pool.py#L181) goes to get the connection and call it's [arequest method](https://github.com/encode/httpcore/blob/ed76a30277529c5687cbdb6e203d6dbd0d56833f/httpcore/_async/connection_pool.py#L203-L220). These connections are set up in [AsyncConnectionPool._create_connection](https://github.com/encode/httpcore/blob/ed76a30277529c5687cbdb6e203d6dbd0d56833f/httpcore/_async/connection_pool.py#L167-L179) and is hardcoded to use the `AsyncHTTPConnection` class. So I need some way to hook into the connection chosen by the transport `arequest()` function but I don't see a way that I can do that without touching private members.

The 3 ways I can think of achieving this, albiet they all use private classes, functions, attributes

### Override _create_connection() and smuggle in my own context and arequest method

```python
class WinRMTransport(httpcore.AsyncConnectionPool):

    def _create_connection(self, *args, **kwargs):
        connection = super()._create_connection(*args, **kwargs)

        context = build_new_context(my_credential_info)
        setattr(connection, '_sec_context', context)
        setattr(connection, '_arequest_old', connection.arequest)

        async def new_arequest(self, method, url, headers, stream, ext):
              context = self._sec_context
              stream = context.wrap(stream)

              return await self._arequest_old(method, url, headers, stream, ext)

        return connection
```

Downsides with this approach

* I'm overriding a private function `_create_connection()`
* I'm smuggling my context in the existing connection object
* I'm overriding `arequest` in a weird and unusual way

### Subclass AsyncHTTPConnection

```python
# Obviously this would need to be publicly exposed somehow
class WinRMConnection(httpcore._async.connection.AsyncHTTPConnection):

    def __init__(self, username, password, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self._sec_context = MyContext(username, password)

    async def arequest(self, method, url, headers, stream, ext):
        stream = self._sec_context.wrap(stream)

        return await super().aqrequest(method, url, headers, stream, ext)


class WinRMTransport(httpcore.AsyncConnectionPool):

    def __init__(self, username, password, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self._username = username
        self._password = password

    def _create_connection(self, origin):
        return WinRMConnection(
            origin=origin,
            http2=self._http2,
            uds=self._uds,
            ssl_context=self._ssl_context,
            local_address=self._local_address,
            retries=self._retries,
            backend=self._backend,

            # Now pass in the args used by WinRMConnection to create my context.
            username=self._username,
            password=self._password,
        )
```

Downsides with this approach

* I'm overriding a private function `_create_connection()`
* To create the connection I need to pass in private attributes and I have no idea if AsyncHTTPConnection will ever add more attributes as time goes on

I personally like this approach the best as it keeps the internal logic largely the same it just need to be solidified in a public interface somehow.


### Reimplement arequest on the transport

```python
class WinRMTransport(httpcore.AsyncConnectionPool):

    def __init__(self, username, password, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self._contexts = {}
        self._username = username
        self._password = password

    async def arequest(self, method, url, headers, stream, ext):
        # Copied from AsyncHTTPTransport with minor tweak
        if url[0] not in (b"http", b"https"):
            scheme = url[0].decode("latin-1")
            raise UnsupportedProtocol(f"Unsupported URL protocol {scheme!r}")
        if not url[1]:
            raise LocalProtocolError("Missing hostname in URL.")

        origin = url_to_origin(url)
        ext = {} if ext is None else ext
        timeout = cast(TimeoutDict, ext.get("timeout", {}))

        await self._keepalive_sweep()

        connection: Optional[AsyncHTTPConnection] = None
        while connection is None:
            async with self._connection_acquiry_lock:
                # We get-or-create a connection as an atomic operation, to ensure
                # that HTTP/2 requests issued in close concurrency will end up
                # on the same connection.
                logger.trace("get_connection_from_pool=%r", origin)
                connection = await self._get_connection_from_pool(origin)

                if connection is None:
                    connection = self._create_connection(origin=origin)
                    logger.trace("created connection=%r", connection)

                    # Create our next context and store one for that connection
                    self._contexts[connection] = NewSecContext(self._username, self._password)

                    await self._add_to_pool(connection, timeout=timeout)
                else:
                    logger.trace("reuse connection=%r", connection)

            context = self._contexts[connection]
            try:
                stream = context.wrap(stream)

                response = await connection.arequest(
                    method, url, headers=headers, stream=stream, ext=ext
                )
            except NewConnectionRequired:
                connection = None
            except Exception:  # noqa: PIE786
                logger.trace("remove from pool connection=%r", connection)
                await self._remove_from_pool(connection)
                raise

        status_code, headers, stream, ext = response
        stream = context.unwrap(stream)

        wrapped_stream = ResponseByteStream(
            stream, connection=connection, callback=self._response_closed
        )
        return status_code, headers, wrapped_stream, ext
```

Downsides with this approach

* I've copied a large block of code which could change at any point in the future making it hard to keep things in sync
* The code references a bunch of private methods and attributes
* Not the biggest fan of using a dict that's keyed to the connection object itself

**Reply by jborean93** [reply]
So after thinking about this a bit more I think the simplest solution is to not use a connection pool and to instantiate my httpx.Client for every task I know will run in parallel with the main connection. This ensures that the auth context is purely for that connection as there will only be one per client. It's more work on the users of the httpx (Async)Client but it's less brittle than drilling down into private interfaces and overriding some of the stuff I mentioned above.

Appreciate everything you've shared so far and I'm happy to answer any questions you have for me.

**Reply by jborean93** [reply]
This brings to my next problem that somewhat ties into this issue. When a connection has gone past the `Keep-Alive` expiry the connection pool will drop the connection and create a new one which invalidates the security context that was for the original connection. Unfortunately there does not seem to be any indicator that this has happened and the connection is dropped after I've already prepared my request. I've got a hack which aligns with one of the options above to hijack into the `aclose()` method on the new connection but it relies on `_create_connection()`.

The [code I'm currently using is](https://github.com/jborean93/pypsrp/blob/21adb94bfb7d7de04ced2ba72e3ad782dfcbe8a3/psrp/io/wsman.py#L33-L45):

```python
class _AsyncWinRMTransport(httpcore.AsyncConnectionPool):

    def _create_connection(self, *args, **kwargs):
        # We need to hook into the connection close() method so we can be alerted when we've started a new connection
        # due to a keep-alive timeout.
        connection = super()._create_connection(*args, **kwargs)
        orig_close = connection.aclose

        async def new_close():
            await orig_close()
            raise _NoAuthenticationContext()
        connection.aclose = new_close
        return connection
```

When the connection pool closes and connections that were timed out this will still close the connection but it raises a custom Exception that lets me know this occured and I need to re set up my authentication context and rewrap my data for the next request. Can you see any other way this might be achievable?

**Reply by jborean93** [reply]
Ok hopefully final update/question I have. I've think I've found a solution to the keep-alive problem, I can check if the response is a 401 which would indicate a new connection was opened and the authentication has not happened. It's not perfect but it's workable and I don't need to touch any of the private methods to achieve it so I'm going for that.

One problem I have is that I cannot seem to indicate the connection is now in an IDLE state after receiving a response so subsequent requests will either use a new connection or hang because the pool is full of "active" connections. I tried doing

```python
# self._transport is `httpcore.AsyncConnectionPool`
response = await self._transport.arequest(method, url, headers, stream, ext)
await response[2].aclose()
```

Which removes the active connection from the pool but instead of making it IDLE so the subsequent request uses that connection it's closed breaking the authentication process. How do I mark a response is read and the connection is free to service a new request?

**Reply by florimondmanca** [reply]
@jborean93 To move the connection to an IDLE state rather than close it, I think you'd need to read through the stream, rather than `.aclose()` it:

```python
response = await self._transport.arequest(method, url, headers, stream, ext)

# Read response body to terminate request-response cycle and allow reusing the connection.
_, _, stream, _ = response
async for _ in stream:
    pass
```

**Reply by jborean93** [reply]
Turns out I had 2 issues

* The initial 401 I get back when sending no auth has the `Connection: close` header which closed the connection so it always created a new one
* You need to both read and close the stream to turn it into `IDLE`

Thanks for all your help.
